src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ChaCha20Poly1305.Android.cs

```diff
@@ -0,0 +1,159 @@
+
+                if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, ciphertextAndTag, out int ciphertextBytesWritten, plaintext))
+                {
+                    throw new CryptographicException();
+                }
+
+                if (!Interop.Crypto.EvpCipherFinalEx(
+                    _ctxHandle,
+                    ciphertextAndTag.Slice(ciphertextBytesWritten),
+                    out int bytesWritten))
+                {
+                    throw new CryptographicException();
+                }
+
+                ciphertextBytesWritten += bytesWritten;
+
+                // NOTE: Android appends tag to the end of the ciphertext in case of ChaCha20Poly1305 and "encryption" mode
+
+                if (ciphertextBytesWritten != ciphertextAndTagLength)
+                {
+                    Debug.Fail($"ChaCha20Poly1305 encrypt wrote {ciphertextBytesWritten} of {ciphertextAndTagLength} bytes.");
+                    throw new CryptographicException();
+                }
+
+                ciphertextAndTag.Slice(0, ciphertext.Length).CopyTo(ciphertext);
+                ciphertextAndTag.Slice(ciphertext.Length).CopyTo(tag);
+            }
+            finally
+            {
+                if (rented is not null)
+                {
+                    CryptoPool.Return(rented, clearSize: ciphertextAndTagLength);
+                }
+            }
+        }
+
+        private void DecryptCore(
+            ReadOnlySpan<byte> nonce,
+            ReadOnlySpan<byte> ciphertext,
+            ReadOnlySpan<byte> tag,
+            Span<byte> plaintext,
+            ReadOnlySpan<byte> associatedData)
+        {
+            Interop.Crypto.EvpCipherSetKeyAndIV(
+                _ctxHandle,
+                ReadOnlySpan<byte>.Empty,
+                nonce,
+                Interop.Crypto.EvpCipherDirection.Decrypt);
+
+            if (!associatedData.IsEmpty)
+            {
+                Interop.Crypto.CipherUpdateAAD(_ctxHandle, associatedData);
+            }
+
+            if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, plaintext, out int plaintextBytesWritten, ciphertext))
+            {
+                CryptographicOperations.ZeroMemory(plaintext);
+                throw new CryptographicException();
+            }
+
+            if (!Interop.Crypto.EvpCipherUpdate(_ctxHandle, plaintext.Slice(plaintextBytesWritten), out int bytesWritten, tag))
+            {
+                CryptographicOperations.ZeroMemory(plaintext);
+                throw new CryptographicException();
+            }
+
+            plaintextBytesWritten += bytesWritten;
+
+            if (!Interop.Crypto.EvpCipherFinalEx(
+                _ctxHandle,
+                plaintext.Slice(plaintextBytesWritten),
+                out bytesWritten))
+            {
+                CryptographicOperations.ZeroMemory(plaintext);
+                throw new CryptographicException(SR.Cryptography_AuthTagMismatch);
+            }
```

Comment:
This order of operations makes no sense to me.  But the tests say it works... so.. OK?

Reply:
This is mostly modeled off of how the GCM implementation does it, but this is my understanding for decrypt

There are two calls to `EvpCipherUpdate` because Android's APIs accept the authentication tag appended to the ciphertext  (`ciphertext || tag`). Android doesn't know how long your plaintext is, so the first update doesn't "decrypt" the number of bytes it was passed, it needs to wait until `final` is called to say, "okay, I've sent all of the ciphertext including the tag".

So the first call sends all of the ciphertext. It may not write the last 16 bytes because it doesn't know that it has reached the tag.

Next, we write the tag. The can also write to the plaintext buffer since the API figured out, "oh wait, you sent me more data, so that data I was hanging on to incase it was the tag can't be the tag, so write it".

Then we call final and that wraps everything up.

Reply:
Oh, right, "input" is last.  "Here's the AD", "here's the ciphertext", "here's the tag", "I'm done".  Seems obvious, just that it's out/in not in/out.